addons/binding/org.openhab.binding.solaredge/src/main/java/org/openhab/binding/solaredge/internal/callback/AbstractCommandCallback.java

```diff
@@ -0,0 +1,182 @@
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.http.HttpStatus.Code;
+import org.openhab.binding.solaredge.config.SolarEdgeConfiguration;
+import org.openhab.binding.solaredge.internal.command.SolarEdgeCommand;
+import org.openhab.binding.solaredge.internal.connector.CommunicationStatus;
+import org.openhab.binding.solaredge.internal.connector.StatusUpdateListener;
+import org.openhab.binding.solaredge.internal.model.DataResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * base class for all commands. common logic should be implemented here
+ *
+ * @author Alexander Friese - initial contribution
+ *
+ */
+public abstract class AbstractCommandCallback extends BufferingResponseListener implements SolarEdgeCommand {
+
+    /**
+     * logger
+     */
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * the configuration
+     */
+    protected final SolarEdgeConfiguration config;
+
+    /**
+     * status code of fulfilled request
+     */
+    private CommunicationStatus communicationStatus;
+
+    /**
+     * listener to provide updates to the WebInterface class
+     */
+    private StatusUpdateListener listener;
+
+    /**
+     * the constructor
+     *
+     * @param config
+     */
+    public AbstractCommandCallback(SolarEdgeConfiguration config) {
+        this.communicationStatus = new CommunicationStatus();
+        this.config = config;
+    }
+
+    /**
+     * the constructor
+     *
+     * @param config
+     */
+    public AbstractCommandCallback(SolarEdgeConfiguration config, StatusUpdateListener listener) {
+        this(config);
+        this.listener = listener;
+    }
+
+    /**
+     * Log request success
+     */
+    @Override
+    public final void onSuccess(Response response) {
+        super.onSuccess(response);
+        communicationStatus.setHttpCode(HttpStatus.getCode(response.getStatus()));
+        logger.debug("HTTP response {}", response.getStatus());
+    }
+
+    /**
+     * Log request failure
+     */
+    @Override
+    public final void onFailure(Response response, Throwable failure) {
+        super.onFailure(response, failure);
+        try {
+            logger.warn("Request failed: {}", failure.toString());
+            communicationStatus.setHttpCode(Code.INTERNAL_SERVER_ERROR);
+
+            // as we are not allowed to catch Throwables we must only throw Exceptions!
+            if (failure instanceof Exception) {
+                communicationStatus.setError((Exception) failure);
+                throw (Exception) failure;
+            }
+        } catch (SocketTimeoutException | TimeoutException e) {
+            communicationStatus.setHttpCode(Code.REQUEST_TIMEOUT);
+        } catch (UnknownHostException e) {
+            communicationStatus.setHttpCode(Code.BAD_GATEWAY);
+        } catch (Exception e) {
+            communicationStatus.setHttpCode(Code.INTERNAL_SERVER_ERROR);
```

Comment:
Are you (re)throwing exceptions to catch them in the same method, that feels, ... non-standard? What is going on here, what are the alternatives?

Reply:
Looks weird.

I think this one is better and doing the same job:

        super.onFailure(response, failure);
        logger.warn("Request failed: {}", failure.toString());
        communicationStatus.setError((Exception) failure);

        if (failure instanceof SocketTimeoutException || failure instanceof TimeoutException) {
            communicationStatus.setHttpCode(Code.REQUEST_TIMEOUT);
        } else if (failure instanceof UnknownHostException) {
            communicationStatus.setHttpCode(Code.BAD_GATEWAY);
        } else {
            communicationStatus.setHttpCode(Code.INTERNAL_SERVER_ERROR);
        }